### PR TITLE
[KOGITO-3858] [KOGITO-4329] [KOGITO-4330] Error Handling

### PR DESCRIPTION
**Pull request by tiagodolphine**
Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [ ] You have read the [contributors guide](CONTRIBUTING.md)
- [ ] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [ ] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains link to any dependent or related Pull Request
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket

### CODE REVIEW

File: jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/FaultNodeVisitor.java [hunk 1]
```diff
@@ -44,8 +46,9 @@ public void visitNode(String factoryField, FaultNode node, BlockStmt body, Varia
             body.addStatement(getFactoryMethod(getNodeId(node), METHOD_FAULT_NAME, new StringLiteralExpr(node.getFaultName())));
         }
 
+        body.addStatement(getFactoryMethod(getNodeId(node), METHOD_TERMINATE_PARENT, new BooleanLiteralExpr(node.isTerminateParent())));
```
Comment (tsurdilo) [inline jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/FaultNodeVisitor.java:49]: this is set for the interrupting flag, correct?

File: kogito-codegen/src/test/resources/error/EndErrorWithEventSubprocess.bpmn2 [hunk 2]
```diff
@@ -0,0 +1,171 @@
+<bpmn2:definitions xmlns:bpmn2="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:bpsim="http://www.bpsim.org/schemas/1.0" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:drools="http://www.jboss.org/drools" id="_IZt5MEe0EDmaKZ1-lMcQzQ" exporter="jBPM Process Modeler" exporterVersion="2.0" targetNamespace="http://www.omg.org/bpmn20">
+  <bpmn2:error id="error1" errorCode="error1"/>
```
Comment (tsurdilo) [inline kogito-codegen/src/test/resources/error/EndErrorWithEventSubprocess.bpmn2:2]: wonder why 2 error defs with same id got generated here

File: jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/ProcessVisitor.java [hunk 3]
```diff
@@ -282,4 +291,23 @@ private void visitCompensationScope(Process process, BlockStmt body) {
             }
         }
     }
+
+    private void visitExceptionScope(Process process, BlockStmt body) {
+        Context context = ((org.jbpm.workflow.core.WorkflowProcess) process).getDefaultContext(CompensationScope.EXCEPTION_SCOPE);
```
Comment (tsurdilo) [inline jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/ProcessVisitor.java:296]: Should this not be gotten from ExceptionScope.EXCEPTION_SCOPE?

File: jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/FaultNodeVisitor.java (hunk 1 above)
Comment (tiagodolphine) [reply jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/FaultNodeVisitor.java:49]: yes, it was missing this property to be set in the codegen, just add the call to the factory.

File: kogito-codegen/src/test/resources/error/EndErrorWithEventSubprocess.bpmn2 (hunk 2 above)
Comment (tiagodolphine) [reply kogito-codegen/src/test/resources/error/EndErrorWithEventSubprocess.bpmn2:2]: don't know, need to check, the erroCode is fine I think but the id I'm not sure

File: jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/ProcessVisitor.java (hunk 3 above)
Comment (tiagodolphine) [reply jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/ProcessVisitor.java:296]: right, my bad

File: kogito-codegen/src/test/resources/error/EndErrorWithEventSubprocess.bpmn2 (hunk 2 above)
Comment (tsurdilo) [reply kogito-codegen/src/test/resources/error/EndErrorWithEventSubprocess.bpmn2:2]: probably in editor it gets created twice once for your error end event and the other for the error start event. its fine maybe we could raise a jira for parser to check if the error def exists. its not a big deal i think as the id and errorcode are same

**Comment by tsurdilo** [comment]
@tiagodolphine just left some questions.  i think this is great and gives a good start for rest of error nodes. thank you!

File: kogito-codegen/src/test/resources/error/EndErrorWithEventSubprocess.bpmn2 (hunk 2 above)
Comment (tiagodolphine) [reply kogito-codegen/src/test/resources/error/EndErrorWithEventSubprocess.bpmn2:2]: @tsurdilo it seems to be generated in this way in the editor, the id for the error definition is the same as the error code.

File: jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/FaultNodeVisitor.java (hunk 1 above)
Comment (tsurdilo) [reply jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/FaultNodeVisitor.java:49]: ok i think would need to be looked at later for boundary error nodes as then can be non interrupting as well. wdyt?

Comment (tiagodolphine) [reply jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/FaultNodeVisitor.java:49]: in fact, this part is the Visitor used in the codegen that basically rebuilt the node using the factory, so in this case it sets the same property terminateParent already set in the Node... but this was set during in the https://github.com/kiegroup/kogito-runtimes/blob/f485259eb4f32bf17afff8475768a2f636b49eea/jbpm/jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/EndEventHandler.java#L85, so if we need to change logic would be in the handler... but this basically parsing the bpmn...

File: kogito-codegen/src/test/resources/error/EndErrorWithEventSubprocess.bpmn2 (hunk 2 above)
Comment (tsurdilo) [reply kogito-codegen/src/test/resources/error/EndErrorWithEventSubprocess.bpmn2:2]: @tiagobento yeah. i think for this pr you could remove one if you wanted to (in both bpmn2 defs used in your tests) but this is just a nitpick :)

File: jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/FaultNodeVisitor.java (hunk 1 above)
Comment (tiagodolphine) [reply jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/FaultNodeVisitor.java:49]: the strange part I needed to set was here https://github.com/tiagodolphine/kogito-runtimes/blob/26a713ab6618dd882f820627db56b17f1f21794f/jbpm/jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/StartEventHandler.java#L156 maybe we are missing some point to set the interrupting property properly, or should it come from bpmn?

File: jbpm/jbpm-bpmn2/src/test/java/org/jbpm/bpmn2/ErrorEventTest.java [hunk 4]
```diff
@@ -499,6 +499,33 @@ public void testBoundaryErrorEventStructureRef() throws Exception {
         assertNodeTriggered(processInstance.getStringId(), "Start", "User Task", "MyBoundaryErrorEvent");
     }
 
+    @Test
+    public void testEndErrorWithSubprocess() throws Exception {
+        KieBase kbase = createKnowledgeBase("error/EndErrorWithEventSubprocess.bpmn2");
```
Comment (cristianonicolai) [inline jbpm/jbpm-bpmn2/src/test/java/org/jbpm/bpmn2/ErrorEventTest.java:504]: check the pr from Daniele to see if we're in sync about how to implement these: https://github.com/kiegroup/kogito-runtimes/pull/1038

**Review by cristianonicolai — COMMENTED** [review]
@tiagodolphine overall looks good! Do you think it is worth adding a process with error handling, into integration tests as well?

File: jbpm/jbpm-bpmn2/src/test/java/org/jbpm/bpmn2/ErrorEventTest.java (hunk 4 above)
Comment (tiagodolphine) [reply jbpm/jbpm-bpmn2/src/test/java/org/jbpm/bpmn2/ErrorEventTest.java:504]: Yeah, once the PR refactoring tests in jbpm is merged I need to fix the tests here as well, but I need that PR to get merged before syncing with the changes that were made.

**Comment by tiagodolphine** [comment]
> @tiagodolphine overall looks good! Do you think it is worth adding a process with error handling, into integration tests as well?

Well don't know, because from the REST API there is nothing that has changed, so I don't know what could I test there that is not already tested in the codegen, wdyt?

**Comment by cristianonicolai** [comment]
@tiagodolphine I was thinking more of a use case point of view, an IT test covering the error handling part from end to end. But that's ok if think is not necessary or include in another PR, just feel that we had some good stability by having major code paths covered in IT tests.

File: jbpm/jbpm-bpmn2/src/test/java/org/jbpm/bpmn2/ErrorEventTest.java (hunk 4 above)
Comment (danielezonca) [reply jbpm/jbpm-bpmn2/src/test/java/org/jbpm/bpmn2/ErrorEventTest.java:504]: The PR has been merged so you can rebase from master :)

Comment (tiagodolphine) [reply jbpm/jbpm-bpmn2/src/test/java/org/jbpm/bpmn2/ErrorEventTest.java]: done

**Review by cristianonicolai — APPROVED** [review]
@tiagodolphine looks great!

**Comment by radtriste** [comment]
Jenkins retest this

**Comment by tiagodolphine** [comment]
jenkins retest this

**Comment by tiagodolphine** [comment]
@MarianMacik in case you have time to review this one, it would be great to merge this ASAP.
Thanks.

**Comment by tiagodolphine** [comment]
jenkins retest this

**Comment by tiagodolphine** [comment]
jenkins retest this

**Review by MarianMacik — APPROVED** [review]

File: jbpm/jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/StartEventHandler.java [hunk 5]
```diff
@@ -149,8 +152,7 @@ protected void handleNode( final Node node, final Element element, final String
                     // BPMN2 spec (p.245-246, (2011-01-03)) implies that
                     //   - a <startEvent> in an Event Sub-Process
                     //    - *without* the 'isInterupting' attribute always interrupts (containing process)
-                    String errorMsg = "Error Start Events in an Event Sub-Process always interrupt the containing (sub)process(es).";
-                    throw new IllegalArgumentException(errorMsg);
+                    startNode.setInterrupting(true);
                 }
```
[the remark this answers is not in the document]
Comment (tiagodolphine) [reply jbpm/jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/StartEventHandler.java]: +1, done

**Comment by tiagodolphine** [comment]
jenkins retest this

**Comment by tiagodolphine** [comment]
jenkins retest this

**Comment by tiagodolphine** [comment]
jenkins retest this

**Comment by tiagodolphine** [comment]
jenkins retest this

**Comment by tiagodolphine** [comment]
jenkins retest this

**Comment by danielezonca** [comment]
jenkins retest this please

**Comment by tiagodolphine** [comment]
jenkins retest this

**Comment by tiagodolphine** [comment]
jenkins retest this

**Comment by tiagodolphine** [comment]
jenkins retest this

**Pull request merged**